### ISSUE TITLE
Add Issue Templates

### ISSUE THREAD

**Issue by yugaego** [company-mode/company-mode#1134]
@dgutov ,

I see that typical questions/answers appear now and then in the issues list.

Maybe, to make the process a bit more efficient,
you would support the idea of adding a template for the bug reports?

I've outlined **roughly** the possible template main points  - and ready to improve if you support the notion - as follows:

1. Lookup in repo issues.
2. Check `M-x company-diag` output.
3. If this is a third-party backend
   - recommend to ask a provider of the backend.
   - link to the Wiki page with the list of the third-party backends.
4. If not
   - ask post `M-x company-diag` output
   - ask post backtrace
     - enable backtracing: `M-x toggle-debug-on-error` **before** the error happens.
     - disable backtracing

What do you think?

[I don't have any ideas for the Feature Request template.
The ideas I have for the _HowTo_ type ones could be incorporated into the home page I think,
again, if you'd like the idea.]

[Github Templates docs](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/about-issue-and-pull-request-templates).

**Comment by dgutov** [comment]
Hey, thanks for taking the initiative!

> you would support the idea of adding a template for the bug reports?

Of course. The recent flow of reports has slowed down compared to, say, a few years ago, so I never got around to this.

> Lookup in repo issues

Good advice.

> If this is a third-party backend

It could also use some note that if backend is `company-capf`, the user should look at the value of `c-a-p-f` printed in the output buffer, and decide accordingly. Perhaps even tell them how to navigate to that function's definition? Which should make clear where it resides.

Or implement the package-finding logic inside `company-diag` itself..?

> enable backtracing: M-x toggle-debug-on-error before the error happens.

Very useful, yes.

> I don't have any ideas for the Feature Request template.

Maybe it could say that the request should describe a usage scenario step-by-step. But we generally get that already.

> The ideas I have for the HowTo type ones could be incorporated into the home page I think, again, if you'd like the idea.

Yes, probably, but could you elaborate on that?

**Comment by yugaego** [comment]
Hey Dmitry,
I started [in a tmp repo](https://github.com/yugaego/tmp/issues/new/choose), first with the classic titles (Bug Report, Feature Request), but then got blown away by creativity winds. Let me know which type of the titles would be a better suit for Company.

All those templates normally go to not gitignored `.github/ISSUE_TEMPLATE` directory.
[eglot](https://github.com/joaotavora/eglot) has it too, so supposedly that shouldn't be a problem for `(M)ELPA`, right? 

If that's **overall** OK, as the next step I plan to create a PR with the templates for the review.

> Perhaps even tell them how to navigate to that function's definition?

Did you mean to mention `C-h f` with a cursor on the function name?

> Or implement the package-finding logic inside `company-diag` itself..?

Did you mean something in the spirit of `(help-fns-function-description-header function)`?
[I didn't figure in which form to pass a `function` to it.]

>> Feature Request template.

> Maybe it could say that the request should describe a usage scenario step-by-step.

[Added](https://github.com/yugaego/tmp/issues/new?assignees=&labels=&template=02-i-ve-got-an-idea.md&title=) based on some adapted examples. Not sure, maybe that contents is too much.

>> The ideas I have for the HowTo type ones

> Yes, probably, but could you elaborate on that?

They ended up in a [question template](https://github.com/yugaego/tmp/issues/new?assignees=&labels=&template=03-i-have-a-question.md&title=).

**Comment by dgutov** [comment]
Hi Y.E!

> eglot has it too, so supposedly that shouldn't be a problem for (M)ELPA, right?

Let's not worry about this. If anybody minds, they can add corresponding ignore entries to the recipe.

> Did you mean to mention C-h f with a cursor on the function name?

Either that or `M-x find-function RET <function-name> RET`.

> Did you mean something in the spirit of (help-fns-function-description-header function)?

Ideally, something that says exactly which ELPA package the function comes from. That seems harder to implement, so let's leave it until some later revision.

> Added based on some adapted examples. Not sure, maybe that contents is too much.

A little bit. One of the reasons to have an issue template is to railroad the user into providing the necessary information up front. From the perspective, having "Blank Issue" seems counter-productive.

Do we need both "I see an error" and "Bug report"? The latter includes the former but it can also manifest like something else than an error. So it seems like a better choice.

"I'd like to discuss" seems covered by "I have a question". And regarding those, while I like the checklist, perhaps enabling the new "Discussions" feature of GitHub could be the better choice. Can one still be presented with a checklist when creating a discussion?

"I've got an idea" could be more traditionally renamed to "Feature request". Could Discussions also be the better place for those? Not sure.

To sum up, the list could look like:

* Bug report
* Feature request
* Ask a question (-> Discussions)

Discussions can also be handily sorted into categories: https://github.com/joaotavora/eglot/discussions/categories/q-a

I was wary of adding discussions because I'm a bit tired of answering the same typical questions (general Emacs ones especially), but help is certainly welcome.

**Comment by yugaego** [comment]
> Either that or M-x find-function RET <function-name> RET.

Added.

> Do we need both "I see an error" and "Bug report"?

Ah, I wasn't clear enough: those were just an example of different titles.

> having "Blank Issue" seems counter-productive

That depends on exact case. If to purposefully direct all the flow to a specific template(s),
then of course `Blank Issue` should be hidden as much as possible (or removed).
This makes sense having, for example, `Discussions` enabled (eglot's case).  
But if everything goes to the Issues, then a hardly visible link to the `Blank Issue` may irritate/confuse.
[Rust repo](https://github.com/rust-lang/rust/issues/new/choose) without `Discussions` is a good example when `Blank Issues` are equally accepted ones.

> perhaps enabling the new "Discussions" feature of GitHub could be the better choice.

Yeah, I almost decided earlier on suggesting you to enable it.
On a closer look, I dislike one weak thing about `Discussions`:
in short, this would be **one more place** to know/remember/take care about (for everyone),
hence a potential source of confusion, outdated information, etc.

> Discussions can also be handily sorted into categories:

Appears those are pre-defined ones: I've enabled `Discussions` on the [test repo](https://github.com/yugaego/tmp/discussions) without editing anything.

And please take a look at the updated `Issues` too:
I've combined `Questions / Feature Request` (with the first idea calling it `Anything Else` Template)
into one template, so that to have only one point of directing *non-bug-reports* to the `Discussions`.
That `Feature Request` template text wasn't particularly meaningful anyway
and `Discussions` have `Ideas` section enabled by default.
But let me know if you think it'd still be better to keep `Feature Requests` at the `Issues`.

> I was wary of adding discussions because I'm a bit tired of answering the same typical questions (general Emacs ones especially)

I added the text about `Emacs` (and third-party) questions to the the `Questions` template.
[Frankly, I was tempted 😈 to link to the [classical article](http://www.catb.org/~esr/faqs/smart-questions.html).]

It'd be wise to [pin](https://docs.github.com/en/discussions/managing-discussions-for-your-community/managing-discussions-in-your-repository#pinning-a-discussion) a similar text to the `Discussions` too.
[and AFAIK, there're no Templates for `Discussions`.]

I hope it covers all for now.
[I'm going to PR (to fix exact wording, etc) once we've settled on the skeleton/flow of the Issues/Discussions.]

Best!
YE

**Comment by dgutov** [comment]
> Rust repo without Discussions is a good example when Blank Issues are equally accepted ones.

That's a lot of options. But Rust is a bit project, nothing comparable to the cozy atmosphere we have in here.

So I suggest we try a limited list first, and then expand it if we see repeated complaints.

> I've combined Questions / Feature Request (with the first idea calling it Anything Else Template)
into one template, so that to have only one point of directing non-bug-reports to the Discussions.

I think we can make that button lead directly to Discussions, right? Without asking the user to find the link in the pre-filled Issue text instead.

The checklist is nice, though. Would be a shame to miss it. If the Discussions don't have a place for such checklist, we could put it somewhere in the docs, or on the home page.

I'm am of double mind regarding the help-gnu-emacs recommendation. I mean, it's a classical mailing list, but both Reddit and Emacs.SE will give more exhaustive and contemporary answers, and are more likely to have the question already answered (and easy to find) in the first place. Those are proprietary platforms, of course, but we are currently using one as well.

> [Frankly, I was tempted smiling_imp to link to the classical article.]

It's a fine article, but it might seem a little rude to ask someone to read a 20-page document first.

Here's something similar (but shorter) about bug reporting, BTW: https://www.gnu.org/software/emacs/manual/html_node/emacs/Understanding-Bug-Reporting.html

> It'd be wise to pin a similar text to the Discussions too.

I suppose the pin could host the checklist we mentioned previously?

> On a closer look, I dislike one weak thing about Discussions:
in short, this would be one more place to know/remember/take care about (for everyone),
hence a potential source of confusion, outdated information, etc.

Yes, that's right. +1 more place to check, answer questions and keep up to date.

OTOH, the main thing I kind of like about it, is it could become the canonical place for the Q&A information. And the way discussions can attach some additional threads to such questions, aside from the main answer, can add more context (as opposed to a wiki article, where this is decidedly less common and has to be added and organized manually).

Right now we have multiple semi-outdated places for such information, and that is not great.

**Comment by yugaego** [comment]
> I think we can make that button lead directly to Discussions, right?

I'm not sure that's possible. Link me to it if you saw this documented or implemented, so I could take a look.
If that'd be possible, it would be a perfect way to redirect from Issues to Discussions.

>  it's a classical mailing list, but both Reddit and Emacs.SE will give more exhaustive and contemporary answers

The links were added.

[For me, after a year into Emacs, figuring out there was help-gnu-emacs from the start was a significant event. And so far it seems to me Emacs "inner" documentation and community is under-represented and under-valued on the Internet, maybe even in a harmful - for Emacs - way.]

> Those are proprietary platforms, of course, but we are currently using one as well.

Which comes with its own pros and cons ;-)  (For the latter I mean Copilot and similar stuff related to proprietary platforms.)

> seem a little rude to ask someone to read a 20-page document first

Yep, that's right.

> https://www.gnu.org/software/emacs/manual/html_node/emacs/Understanding-Bug-Reporting.html

Yes, a good one.


I added two PRs after considering everything discussed so far.

The suggested approach uses Discussions with a pinned `Before Asking a Question` post.

Though I'm equally fine with using only Issues (Bug Report, Question, Feature Request),
if eventually you'd decided Discussions cons are overweighting (one more minus that came to my mind is that Discussions may cause even stronger locking to the GitHub Platform).

**Comment by dgutov** [comment]
> I'm not sure that's possible. Link me to it if you saw this documented or implemented, so I could take a look.

Yeah, apparently not.

> And so far it seems to me Emacs "inner" documentation and community is under-represented and under-valued on the Internet, maybe even in a harmful - for Emacs - way.

The same "inner" community has arguably expended significant effort to keep things as they are, whether intentionally or unintentionally. I hope this will change over time.

So far I think the best signs for it are Emacs maintainers being active and responsive in Reddit comments.

>  (For the latter I mean Copilot and similar stuff related to proprietary platforms.)

Yeah, I'm not happy about that one.

> if eventually you'd decided Discussions cons are overweighting (one more minus that came to my mind is that Discussions may cause even stronger locking to the GitHub Platform)

We could use it for a while, and then copy all the most best (and nicely organized) questions into some wiki document somewhere.

Another bonus of discussions I have just found out about, is we can convert existing Issues into them. So the Q&A section might require even less up-front work to start being useful.

**Comment by yugaego** [comment]
> we can convert existing Issues into them.

Here's a good recent candidate: https://github.com/company-mode/company-mode/issues/1138

>  has arguably expended significant effort to keep things as they are
> I hope this will change over time.

Even though I heard the (and related) rumors and an overall after-taste I noticed **was** at some moment "better stay away from that community", from my own experience, the Emacs community **is** welcoming enough so far.
So yes, I hope there **is** an improvement in this regard too.  
IMO, it'd be a huge mistake for the future of Emacs to underestimate the role of the whole community and every human in it.

Anyway, I suppose we both would agree that's enough for an off-topic :-), so closing this issue.
Have a nice day!

**Comment by yugaego** [comment]
> But the community is both welcoming and very committed to keeping things as they are. Which invariably results in not having to welcome too many new participants.

Smartly formulated 🥇
